lib/pengine/container.c

```diff
@@ -0,0 +1,800 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include <crm_internal.h>
+
+#include <crm/pengine/rules.h>
+#include <crm/pengine/status.h>
+#include <crm/pengine/internal.h>
+#include <unpack.h>
+#include <crm/msg_xml.h>
+
+#define VARIANT_CONTAINER 1
+#include "./variant.h"
+
+void tuple_free(container_grouping_t *tuple);
+
+static char *
+next_ip(const char *last_ip)
+{
+    int oct1 = 0;
+    int oct2 = 0;
+    int oct3 = 0;
+    int oct4 = 0;
+
+    int rc = sscanf(last_ip, "%d.%d.%d.%d", &oct1, &oct2, &oct3, &oct4);
+    if (rc != 4) {
+        return NULL;
+
+    } else if(oct4 > 255) {
+        return NULL;
+    }
+
+    return crm_strdup_printf("%d.%d.%d.%d", oct1, oct2, oct3, oct4+1);
+}
+
+static int
+allocate_ip(container_variant_data_t *data, container_grouping_t *tuple, char *buffer, int max) 
+{
+    if(data->ip_range_start == NULL) {
+        return 0;
+
+    } else if(data->ip_last) {
+        tuple->ipaddr = next_ip(data->ip_last);
+
+    } else {
+        tuple->ipaddr = strdup(data->ip_range_start);
+    }
+
+    data->ip_last = tuple->ipaddr;
+#if 0
+    return snprintf(buffer, max, " --add-host=%s-%d:%s --link %s-docker-%d:%s-link-%d",
+                    data->prefix, tuple->offset, tuple->ipaddr,
+                    data->prefix, tuple->offset, data->prefix, tuple->offset);
+#else
+    return snprintf(buffer, max, " --add-host=%s-%d:%s",
+                    data->prefix, tuple->offset, tuple->ipaddr);
+#endif
+}
+
+static xmlNode *
+create_resource(const char *name, const char *provider, const char *kind) 
+{
+    xmlNode *rsc = create_xml_node(NULL, XML_CIB_TAG_RESOURCE);
+
+    crm_xml_add(rsc, XML_ATTR_ID, name);
+    crm_xml_add(rsc, XML_AGENT_ATTR_CLASS, "ocf");
+    crm_xml_add(rsc, XML_AGENT_ATTR_PROVIDER, provider);
+    crm_xml_add(rsc, "type", kind);
+
+    return rsc;
+}
+
+static void
+create_nvp(xmlNode *parent, const char *name, const char *value) 
+{
+    char *id = crm_strdup_printf("%s-%s", ID(parent), name);
+    xmlNode *xml_nvp = create_xml_node(parent, XML_CIB_TAG_NVPAIR);
+
+    crm_xml_add(xml_nvp, XML_ATTR_ID, id); free(id);
+    crm_xml_add(xml_nvp, XML_NVPAIR_ATTR_NAME, name);
+    crm_xml_add(xml_nvp, XML_NVPAIR_ATTR_VALUE, value);
+}
+
+static void
+create_op(xmlNode *parent, const char *prefix, const char *task, const char *interval) 
+{
+    char *id = crm_strdup_printf("%s-%s-%s", prefix, task, interval);
+    xmlNode *xml_op = create_xml_node(parent, "op");
+
+    crm_xml_add(xml_op, XML_ATTR_ID, id); free(id);
+    crm_xml_add(xml_op, XML_LRM_ATTR_INTERVAL, interval);
+    crm_xml_add(xml_op, "name", task);
+}
+
+static bool
+create_ip_resource(
+    resource_t *parent, container_variant_data_t *data, container_grouping_t *tuple,
+    pe_working_set_t * data_set) 
+{
+    if(data->ip_range_start) {
+        char *id = crm_strdup_printf("%s-ip-%s", data->prefix, tuple->ipaddr);
```

Comment:
There are a couple of other places in the code base where we "sanitize" XML IDs (currently just replacing ':' and '#' with '.'). It would probably be a good idea to make that a common function and use it here as well, so this would work with IPv6 addresses.